reasoner/resolution/resolver/BoundConclusionResolver.java

```diff
@@ -0,0 +1,377 @@
+            sendAnswerOrSearchDownstreamOrFail(w.first(), requestState, w.second());
+        });
+    }
+
+    protected Pair<Request, Integer> fromUpstream(Materialiser.Request toDownstream) {
+        assert materialiserRequestRouter.containsKey(toDownstream);
+        return materialiserRequestRouter.remove(toDownstream);
+    }
+
+    private void sendAnswerOrSearchDownstreamOrFail(Request fromUpstream, ConclusionRequestState<?> requestState, int iteration) {
+        Optional<? extends AnswerState.Partial<?>> upstreamAnswer = requestState.nextAnswer();
+        if (upstreamAnswer.isPresent()) {
+            answerToUpstream(upstreamAnswer.get(), fromUpstream, iteration);
+        } else if (!requestState.isComplete() && requestState.downstreamManager().hasDownstream()) {
+            requestFromDownstream(requestState.downstreamManager().nextDownstream(), fromUpstream, iteration);
+        } else if (requestState.waitedMaterialisations().waiting()) {
+            requestState.waitedMaterialisations().addWaitingRequest(fromUpstream, iteration);
+        } else {
+            requestState.setComplete();
+            failToUpstream(fromUpstream, iteration);
+        }
+    }
+
+    @Override
+    protected void receiveFail(Response.Fail fromDownstream, int iteration) {
+        LOG.trace("{}: received Fail: {}", name(), fromDownstream);
+        if (isTerminated()) return;
+
+        Request toDownstream = fromDownstream.sourceRequest();
+        Request fromUpstream = fromUpstream(toDownstream);
+        ConclusionRequestState<?> requestState = this.requestStates.get(fromUpstream);
+
+        if (iteration < requestState.iteration()) {
+            // short circuit old iteration fail messages to upstream
+            failToUpstream(fromUpstream, iteration);
+            return;
+        }
+
+        requestState.downstreamManager().removeDownstream(fromDownstream.sourceRequest());
+        sendAnswerOrSearchDownstreamOrFail(fromUpstream, requestState, iteration);
+    }
+
+    @Override
+    protected void initialiseDownstreamResolvers() {
+        throw TypeDBException.of(ILLEGAL_STATE);
+    }
+
+    private ConclusionRequestState<?> createRequestState(Request fromUpstream, int iteration) {
+        LOG.debug("{}: Creating a new ConclusionResponse for request: {}", name(), fromUpstream);
+
+        ConclusionRequestState<?> requestState;
+        if (fromUpstream.partialAnswer().asConclusion().isExplain()) {
+            requestState = new ConclusionRequestState.Explain(fromUpstream, iteration, conditionDownstreams(fromUpstream));
+        } else if (fromUpstream.partialAnswer().asConclusion().isMatch()) {
+            requestState = new ConclusionRequestState.Match(fromUpstream, iteration, conditionDownstreams(fromUpstream));
+        } else {
+            throw TypeDBException.of(ILLEGAL_STATE);
+        }
+
+        return requestState;
+    }
+
+    private List<Request> conditionDownstreams(Request fromUpstream) {
```

Comment:
it's one downstream but multiple possible "completions" given a concept that is meant to be generated! These completions are all for the same `Rule.Condition` (since this conclusion belongs to just 1 rule), but all may be different in terms of bounds

Reply:
Yes! Is this a suggestion for a change?